### PR TITLE
Better connection management

### PR DESCRIPTION
**Pull request by piodul**
Reworks how the driver manages connections per-node connections.

A new object, `NodeConnectionPool` is introduced which takes responsibility of managing connections from `Node` and `ConnectionKeeper`. The connection keeper was not removed as it is still used for managing the control connection.

The new management scheme has following advantages over the previous approach:

- The number of connections per node is now configurable.
- Support for establishing per-shard connections over the non-shard-aware port (9042),
- Automatic detection of the shard-aware port on a per-node basis - the driver uses it unless explicitly disabled in the configuration,
- In case the shard-aware port is unreachable or hidden behind source-port translating NAT, the driver falls back to using the non-shard-aware port.

__TODO__: I did some minimal testing (node restart, node add/remove) but should probably do some more (resharding, shard-aware port unreachable, shard-aware port is disabled/enabled after restart). For now, marking as draft.

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [ ] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [ ] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

**Comment by psarna** [comment]
@Jasperav it's under review right now, and it also needs a little cleanup wrt. commits and commit messages. Overall, next week is very likely.

File: scylla/src/transport/connection_pool.rs [hunk 1]
```diff
@@ -0,0 +1,1042 @@
...
+        }
+    }
+}
+
+enum MaybePoolConnections {
+    // The pool is still being initialized
+    Initializing,
+
+    // The pool is empty and is waiting to be refilled
+    Pending,
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/connection_pool.rs]: OK. Changed to `Broken`

File: scylla/src/transport/connection_pool.rs [hunk 2]
```diff
@@ -0,0 +1,1042 @@
...
+    ) -> Arc<Connection> {
+        // Try getting the desired connection
+        if let Some(conn) = Self::choose_random_connection_from_slice(&shard_conns[shard as usize])
+        {
+            return conn;
+        }
+
+        // If this fails try getting any other in random order
+        let mut shards_to_try: Vec<u16> =
+            (shard..nr_shards.get()).chain(0..shard).skip(1).collect();
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/connection_pool.rs]: Using `filter` has the following drawback:
- There will be an additional check to be performed for each shard ID
- The iterator won't be able to provide an accurate [`size_hint`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.size_hint) due to the use of `filter`, so `collect` won't be able to pre-allocate the right amount of memory for the Vec.

How about this instead (this is what I pushed in the updated version):

```rust
let mut shards_to_try: Vec<u16> = (0..shard).chain(shard + 1..nr_shards.get()).collect();
```

**Comment by piodul** [comment]
v2:
- Slightly changed how the `shards_to_try` vector is being populated,
- Renamed `MaybeConnectionPool::Pending` to `MaybeConnectionPool::Broken`,
- Got rid of `ConnectionKeeper`,
- Slightly changed the refilling algorithm: now, if the pool is empty and it starts filling, it always opens exactly one connection. This is done in order to reduce the number of connection attempts while the target node is down.

**Review by havaker — APPROVED** [review]

**Comment by piodul** [comment]
v2.1:
- Reduced verbosity of log messages which appear when pool is being refilled and the shard-aware port is being used
- Corrected the excess pool maximum size calculation - now it's `node_shard_count * 10` (as in gocql), previously it was `per_shard_target * 10`

**Comment by piodul** [comment]
I did a series of manual tests - I used an application which writes with a steady rate, and it calls `use_keyspace` at startup.

Ran this scenario twice, with two pool sizes - `PerHost` and `PerShard`:

- Start with one node
- Add a new node
- Restart one node with less shards
- Restart one node with more shards
- Restart one node with shard-aware port disabled
- Restart one node with shard awareness disabled
- Restart one node, while the node is down restart the application
- Decommission node

Ran this scenario with `PerShard` strategy:

- Deliberately introduce a bug to source-port choosing algorithm, connect to a node and observe how the fallback-to-non-shard-aware-port logic behaves

Ran this scenario with `PerShard` strategy, shard-aware port disallowed:

- Start multiple instances of the application and observe the connection storm which eventually passes

The application behaved in a sane way in all of the scenarios.

**Comment by Jasperav** [comment]
I think it's not to hard to spin up additional nodes with Docker with Command within a Rust test. Maybe add a test which is mentioned above while fixing https://github.com/scylladb/scylla-rust-driver/issues/307

**Comment by psarna** [comment]
Adding a simple distributed test in order to solve #307 is definitely a good idea, but anything more complex is very likely to burn our CI credits very quickly, so for now a set of manual tests is also acceptable. I'll proceed with merging and after that we can start planning a `0.3.0` release, after we browse the statuses of other pending pull requests and move them forward as well.

**Comment by Jasperav** [comment]
@psarna CI credit? Github Action minutes are completely free for public repositories: https://github.com/pricing

**Comment by psarna** [comment]
Isn't there an upper bound of minutes to use though? It would make sense for Github to place some sanely limits in order to prevent abuse. If there isn't, then indeed we can add a bunch of distributed tests.

**Comment by Jasperav** [comment]
I don't think so! Yes! Distributed tests are missing for this crate, that would make it a lot faster production ready :)

**Comment by QuentinPerez** [comment]
so cool thank you :)

**Pull request merged**